[FLEX-623] Code changes for Simple Toast Without Vertical Stacking

payald opened the pull request:
Added Code changes for Toast. Added a fix for modal as well.
The issue that is fixed for modal is that when modal is open and we navigate to another page, and then navigate back to modal page then page doesn't render. To fix this, code has been added to componentWillMount.
<img width="956" alt="toast_done_screenshot_1" src="https://user-images.githubusercontent.com/10673883/61216100-3586c300-a72a-11e9-89e7-c942931cf29a.png">
<img width="958" alt="toast_done_screenshot_2" src="https://user-images.githubusercontent.com/10673883/61216102-3586c300-a72a-11e9-9bfe-64b2ca7c5e66.png">
<img width="957" alt="toast_done_screenshot_3" src="https://user-images.githubusercontent.com/10673883/61216104-3586c300-a72a-11e9-995f-6501aeae2c0c.png">
<img width="957" alt="toast_done_screenshot_4" src="https://user-images.githubusercontent.com/10673883/61216105-361f5980-a72a-11e9-97e4-109cdb6871ba.png">

payald commented:
Please find the screenshots attached.
<img width="954" alt="toast_done_screenshot_1" src="https://user-images.githubusercontent.com/10673883/60969690-c54cfb80-a33d-11e9-953c-1f9a45f15ca6.png">
<img width="956" alt="toast_done_screenshot_2" src="https://user-images.githubusercontent.com/10673883/60969691-c54cfb80-a33d-11e9-8844-09b8b0d1b568.png">

psivakumaran commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+        this.state = {
+            showToast: false,
+            msg: "",
+            type: "success",
```
Please can you add enum constants for "position" and "type"?

payald replied:
Addressed comment

psivakumaran commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+        const toastType = options.type === undefined ? "success" : options.type;
+        const pos = options.position === undefined ? "top-right" : options.position;
+
+        if (this.state.showToast === false) {
```
Should this be vice-versa, I mean default should be true, and when we need to wait till it fades, it should be false.

psivakumaran commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+                            this.setState({
+                                showToast: false,
+                                msg: "",
+                                type: "success",
```
Should this be "toastType" ?

payald replied:
After timeout, I am resetting all the values. Hence type is reset to success

psivakumaran commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+                    hideAfter: hideTime,
+                    style: {
+                        opacity: 1,
+                        [marginTypeForTopOrBottom]: "15px",
```
Please can you add constant for "15px"

payald replied:
Addressed comment

psivakumaran commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+    /* Retrieve relevant icon based on type of message */
+    getIcon() {
+        switch (this.state.type) {
+            case "warning":
```
These cases should come from enum

payald replied:
Addressed comment

payald commented on src/toast/TestMyToaster.tsx:
```diff
@@ -0,0 +1,41 @@
+import * as React from "react";
+import toast from "../toast";
+
+export type ButtonProps = {};
```
Removed Buttonprops

payald commented on src/toast/TestMyToaster.tsx:
```diff
@@ -0,0 +1,41 @@
+
+export type ButtonProps = {};
+
+export class TestMyToaster extends React.Component<ButtonProps> {
```
Addressed comment

payald commented on src/toast/TestMyToaster.tsx:
```diff
@@ -0,0 +1,41 @@
+    render() {
+        return (
+            <div>
+                <button onClick={this.handleClickForBottomLeft}>Show Toaster at bottom left corner</button>
```
Addressed comment

payald commented:
Please find the screenshots attached.
<img width="956" alt="toast_done_screenshot_1" src="https://user-images.githubusercontent.com/10673883/61034435-e582c600-a3e2-11e9-99eb-57ae28f05df8.png">
<img width="957" alt="toast_done_screenshot_2" src="https://user-images.githubusercontent.com/10673883/61034437-e582c600-a3e2-11e9-9faa-12724407b371.png">
<img width="957" alt="toast_done_screenshot_3" src="https://user-images.githubusercontent.com/10673883/61034439-e61b5c80-a3e2-11e9-9405-4ad76adb338b.png">
<img width="957" alt="toast_done_screenshot_4" src="https://user-images.githubusercontent.com/10673883/61034440-e61b5c80-a3e2-11e9-93b5-7ced34242573.png">

payald commented on src/toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,22 @@
+    <div>
+        <h4> Toast </h4>
+
+        <TestMyToaster name="Toast is done" />
```
Addressed comment

payald commented on src/toast/ToastContainer.tsx:
```diff
@@ -0,0 +1,46 @@
+
+const ID = "toast-container";
+
+class ToastContainer {
```
This class is not extended from React.PureComponent because it is being used to access the Toast Component.

payald commented on src/toast/ToastContainer.tsx:
```diff
@@ -0,0 +1,46 @@
+const ID = "toast-container";
+
+class ToastContainer {
+    functor?: Function;
```
Addressed comment

payald commented on src/toast/index.ts:
```diff
@@ -0,0 +1,5 @@
+
+const toast = new ToastContainer();
+
+export default toast;
```
Addressed comment

payald commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+    render() {
+        let classes = `toast ${this.state.type} `;
+        return (
+            <div className={classes} style={this.state.style}>
```
Addressed comment

psivakumaran commented:
> Please find the screenshots attached.
> <img alt="toast_done_screenshot_1" width="956" src="https://user-images.githubusercontent.com/10673883/61034435-e582c600-a3e2-11e9-99eb-57ae28f05df8.png">
> <img alt="toast_done_screenshot_2" width="957" src="https://user-images.githubusercontent.com/10673883/61034437-e582c600-a3e2-11e9-9faa-12724407b371.png">
> <img alt="toast_done_screenshot_3" width="957" src="https://user-images.githubusercontent.com/10673883/61034439-e61b5c80-a3e2-11e9-9405-4ad76adb338b.png">
> <img alt="toast_done_screenshot_4" width="957" src="https://user-images.githubusercontent.com/10673883/61034440-e61b5c80-a3e2-11e9-93b5-7ced34242573.png">

Please can you the space between icon and the message in toast?

payald commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,93 @@
+import {Icon} from "../icon";
+import "./styles.css";
+
+export default class Toast extends React.Component<Props> {
```
Addressed comment.

ben-schumacher replied:
I'm not sure I follow @payald.  You can still utilize state with PureComponent.

payald replied:
Updated code

akate1 commented on src/toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,21 @@
+storiesOf("Toast", module).add("Toast Simple", () => (
+    <div>
+        <h4> Toast </h4>
+        <TestToaster />
```
Here you are suppose to add code which shows how user will use the Toast component in other application.

payald replied:
Addressed this comment.

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+};
+
+interface ToastProps {
+    notify: (fn: Function) => void;
```
You should also take message from  user to show in Toast.

payald replied:
Text message is taken as input.

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+        super(props);
+        props.displayToast(this.show_toast);
+
+        this.state = {
```
you can keep this outside of constructor.

payald replied:
Addressed this comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+            msg: "",
+            type: MessageType.SUCCESS,
+            position: Direction.TOP_RIGHT,
+            hideAfter: 3,
```
Please define all your props before using them. and Do not add hard coded values anywhere. Create constants for this or make them default values for props.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+                    type: toastType,
+                    position: pos,
+                    hideAfter: hideTime,
+                    style: {
```
I am still not getting why we need to put style in component state ?

payald replied:
Removed style.

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+                () => {
+                    setTimeout(
+                        () =>
+                            this.setState({
```
If you need to set state at many places write one function to set component state instant of repeating code.

payald replied:
Updated code

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,153 @@
+    };
+    render() {
+        return (
+            <div className={`toast ${this.state.type}`} style={this.state.style}>
```
is 'toast' here is classname ? if yes then the code should be {` ${"toast"} ${ this.state.type} `}

payald replied:
Updated code

akate1 commented on src/toast/ToastContainer.tsx:
```diff
@@ -0,0 +1,47 @@
+
+const ID = "toast-container";
+
+class ToastContainer {
```
I think you can club this ToastContainer class code in 'Toast' class itself. Please refer code for 'Modal' component.

ben-schumacher replied:
I don't think this class is needed at all and we need to follow standard React programming style.  Any class we use should implement React component.  We also should not be accessing the DOM directly and instead utilize React methods.

payald replied:
Updated code.

payald commented on src/toast/TestToaster.tsx:
```diff
@@ -0,0 +1,66 @@
+                    <Button onClick={this.handleClickForTopRight}>Show Toaster at top right corner</Button>
+                    <Button onClick={this.handleClickForBottomRight}>Show Toaster at bottom right corner</Button>
+                    <Button onClick={this.handleClickForTopLeft}>Show Toaster at top left corner</Button>
+                    <Button onClick={this.handleClickWithMandatoryParams}>Show Toaster with mandatory params</Button>
```
Removed mandatory params

payald commented on src/toast/TestToaster.tsx:
```diff
@@ -0,0 +1,66 @@
+
+export type TestToasterProps = {};
+
+export class TestToaster extends React.PureComponent<TestToasterProps> {
```
Removed this file

payald commented on src/toast/TestToaster.tsx:
```diff
@@ -0,0 +1,66 @@
+            <div className="clr-row">
+                <div className="clr-col-12">
+                    <Button onClick={this.handleClickForBottomLeft}>Show Toaster at bottom left corner</Button>
+                    <Button onClick={this.handleClickForTopRight}>Show Toaster at top right corner</Button>
```
Addressed comment

payald commented on src/toast/TestToaster.tsx:
```diff
@@ -0,0 +1,66 @@
+        return (
+            <div className="clr-row">
+                <div className="clr-col-12">
+                    <Button onClick={this.handleClickForBottomLeft}>Show Toaster at bottom left corner</Button>
```
Removed this file. The stories are added like the one created for modals.

payald commented on src/toast/TestToaster.tsx:
```diff
@@ -0,0 +1,66 @@
+
+export type TestToasterProps = {};
+
+export class TestToaster extends React.PureComponent<TestToasterProps> {
```
Removed this file.

payald commented:
> > Please find the screenshots attached.
> > <img alt="toast_done_screenshot_1" width="956" src="https://user-images.githubusercontent.com/10673883/61034435-e582c600-a3e2-11e9-99eb-57ae28f05df8.png">
> > <img alt="toast_done_screenshot_2" width="957" src="https://user-images.githubusercontent.com/10673883/61034437-e582c600-a3e2-11e9-9faa-12724407b371.png">
> > <img alt="toast_done_screenshot_3" width="957" src="https://user-images.githubusercontent.com/10673883/61034439-e61b5c80-a3e2-11e9-9405-4ad76adb338b.png">
> > <img alt="toast_done_screenshot_4" width="957" src="https://user-images.githubusercontent.com/10673883/61034440-e61b5c80-a3e2-11e9-93b5-7ced34242573.png">
> 
> Please can you the space between icon and the message in toast?

Addressed comment

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+const PADDING_RIGHT = "padding-right";
+
+const initialState = {
+    isToastRendering: false,
```
This is the only value we should need to keep in state.  All other values can be just properties of the Toast component.

payald replied:
Updated code.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+    hideAfter: DEFAULT_HIDE_TIME,
+};
+
+const ICONTYPE = {
```
To be more flexible, can we just have another property `icon: string` that gets passed as the value for the shape on the Icon component?

payald replied:
Updated code.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+                : TOAST_CLASSES.LEFT + " " + TOAST_CLASSES.MARGIN_LEFT;
+            classes = classes + " " + classname;
+        } else {
+            classes = TOAST_CLASSES.INVISIBLE + " ";
```
We should have to style the div as hidden/invisible.  We can just check the state and either render the div into the DOM or completely omit it.  Please check Model for some pointers on how to do this.

payald replied:
Updated code.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+        );
+    }
+    /* Retrieve relevant icon based on type of message */
+    getIcon() {
```
Can remove this if we just pass icon as a property that gets passed as the shape.

payald replied:
Updated code

ben-schumacher commented on src/toast/styles.css:
```diff
@@ -0,0 +1,86 @@
+.toast {
```
Is there any way we can utilize existing Clarity classnames?  For example, there should already be classnames that we use for the background-color in other components.  I know this isn't a standard VMWare Clarity component, but it would be great if we stayed as closely aligned as possible to the current tool set (in terms of styling).

payald replied:
Updated code.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+ */
+
+import * as React from "react";
+import "@clr/icons/clr-icons.min.css";
```
not needed

payald replied:
Updated code

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+
+import * as React from "react";
+import "@clr/icons/clr-icons.min.css";
+import "@webcomponents/custom-elements/custom-elements.min.js";
```
not needed

payald replied:
Updated code

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,163 @@
+import * as React from "react";
+import "@clr/icons/clr-icons.min.css";
+import "@webcomponents/custom-elements/custom-elements.min.js";
+import "@clr/icons/clr-icons.min.js";
```
not needed

payald replied:
Updated code.

payald commented on src/toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,63 @@
+
+import * as React from "react";
+import {storiesOf} from "@storybook/react";
+import {action} from "@storybook/addon-actions";
```
Updated code.

payald commented:
> * Please merge origin/master into this branch.
> * Please use .tsx file extension to stay consistent.
> * Please add Dell copyright to top of all source files.

Updated code.

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+    type?: MessageType;
+    icon?: string;
+    text: string;
+    onClose: Function;
```
This can be optional prop. By default toast will close after specified timeout.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+        type: MessageType.SUCCESS,
+        position: Direction.TOP_RIGHT,
+        duration: DEFAULT_HIDE_TIME,
+        icon: IconType.SUCCESS,
```
Are we sure that we need to pass default Icon ? because there can be requirement of Toast with no icon as well.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+            document.body.appendChild(el);
+            this.divRef = el;
+        }
+        document.body.classList.add(ClassNames.NO_SCROLLING);
```
Don't we need this line inside if statement ?

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+            this.createElement();
+
+            if (nextProps.showToast === true && nextProps.showToast !== nextState.showToast) {
+                let duration = nextProps.duration === undefined ? DEFAULT_HIDE_TIME : nextProps.duration;
```
No need of this I think default prop will take care of assigning default values if that is not pass by props.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+            document.body.removeChild(this.divRef);
+            this.divRef = null;
+        }
+        document.body.classList.remove(ClassNames.NO_SCROLLING);
```
Check if we need this inside if statement as well.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+
+        return (
+            <div className={classNames(["alert", `alert-${type}`, className])} style={styled}>
+                <span style={{paddingRight: "5px"}}>{icon && <Icon shape={icon} />}</span>
```
Check if user has passed icon then only render this span tag

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,170 @@
+        };
+
+        //create styles based on the position given
+        let place = (position || Direction.TOP_RIGHT).includes(Directions.BOTTOM) ? Directions.BOTTOM : Directions.TOP;
```
you have already set default value for 'position' as Direction.TOP_RIGHT so you can directly check position here. 
Also, please check if we really need to use enum 'Directions' here. You can use constants as well or use strings directly.

payald replied:
Addressed comment

akate1 commented on src/toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,93 @@
+import {State, Store} from "@sambego/storybook-state";
+import {Toast} from "./Toast";
+import {Button} from "../forms/button";
+import {Direction, MessageType} from "./Toast";
```
You can import Toast here no need of two imports from /Toast

payald replied:
Addressed comment

payald commented on src/toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,93 @@
+
+const DURATION = 4000;
+
+enum IconType {
```
Addressed comment

akate1 commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,162 @@
+    INFO = "info",
+}
+
+enum IconType {
```
Please remove this enum as it is not needed now.

ben-schumacher commented on src/modals/Modal.tsx:
```diff
@@ -49,17 +49,25 @@ export class Modal extends React.PureComponent<ModalProps> {
     }
 
+    componentWillMount() {
+        this.createDivRef();
```
Why do we have to create the placeholder <div/> here?  Can't we just rely on componentWillUpdate to create the <div/> placeholder?  I commented out this function and everything seems to work fine.

payald replied:
I was facing an issue where we navigate to another page from the modal page with an opened modal and move back to the modal page I get the error "Target  container is not a DOM element". Arpita had already resolved this issue using componentWillMount().

![image](https://user-images.githubusercontent.com/10673883/61956373-9c359780-afda-11e9-9f7a-f5fac4fc5bff.png)

ben-schumacher replied:
Okay, got it.  Thanks.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    componentWillMount() {
```
Same comment from Modal.tsx.  Why do we need this?  Can't we just rely on componentWillUpdate to create the placeholder <div/>?

payald replied:
I was initially facing similar issue in Toast as well.

ben-schumacher replied:
Okay, got it.  Thanks.

ben-schumacher commented on src/toast/Toast.tsx:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    componentWillUnmount() {
+        this.close();
```
Why do we call this.close() here?

payald replied:
I was facing an issue where if the toast is visible and I navigate to another page and then come back to the toast page then the previous toast is still visible as it was not closed earlier (because component unmounted while navigating to the other page). To fix this I have added this.close() in componentWillUnmount() to close the toast if we are navigating to another page while the toast is visible.

payald replied:
Removed this.close() from componentWillUnmount.

ben-schumacher replied:
Yeah, I think this is the right decision.  It seems a little difficult to cover all use-cases in Storybook environment.  Hopefully, in a true we UI the Toast will be higher up in the component chain a can still display/hide across multiple pages.

ben-schumacher commented:
@payald, don't forget to merge master or rebase.  Thanks!
